File: solutions/20.valid-parentheses.py
#
# @lc app=leetcode id=20 lang=python3
#
# [20] Valid Parentheses
#

# @lc code=start
class Solution:
    def isValid(self, s: str) -> bool:
        pair = {
            '(': ')',
            '{': '}',
            '[': ']',
        }

        stack = []

        for char in s:
            if char in pair:
                stack.append(pair[char])
            elif len(stack) > 0 and stack[-1] == char:
                stack.pop()
            else:
                return False
        return len(stack) == 0
# @lc code=end

File: solutions/20.valid-parentheses.ts
/*
 * @lc app=leetcode id=20 lang=typescript
 *
 * [20] Valid Parentheses
 */

// @lc code=start
function isValid(s: string): boolean {
  const brackets = {
    "(": ")",
    "{": "}",
    "[": "]",
  }

  const stack: string[] = []

  for(let i = 0; i < s.length; i++ ){
    if(brackets[s[i]]) {
      stack.push(brackets[s[i]])
    } else if (stack && s[i] === stack[stack.length - 1]) {
      stack.pop()
    } else {
      return false
    }
  }
  console.log(stack.length)

  return stack.length === 0
};
// @lc code=end
